New module: routeros — manage MikroTik RouterOS

heuels opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
For quite some time I have been seeing questions popping up on StackOverflow, official wiki and even here (#29333) to add native support for managing MikroTik RouterOS devices to Ansible. There have been several projects, but all of them either don't support the latest Ansible release or are long abandoned. This pull request implements a simple `network_cli` module to support RouterOS devices. I also plan to further support this module and add more commands and functionality to it in the future.

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
`routeros_command`
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (routeros 4cd32a8cc6) last updated 2018/06/05 21:37:11 (GMT +300)
  config file = None
  configured module search path = ['/Users/heuels/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/heuels/.pyenv/versions/3.6.5/lib/python3.6/site-packages/ansible/lib/ansible
  executable location = /Users/heuels/.pyenv/versions/3.6.5/lib/python3.6/site-packages/ansible/bin/ansible
  python version = 3.6.5 (default, May  8 2018, 10:48:52) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

heuels committed:
Implement initial RouterOS support

heuels committed:
Correct matchers for license prompts

heuels committed:
Documentation updates & mild refactor

heuels committed:
Remove one last Cisco function

heuels committed:
Sanity test fixes

heuels committed:
Move imports to the beginning

heuels committed:
Remove authorize property

felixfontein commented:
I don't have any experience so far with network modules, so maybe I'm doing something really wrong ;) I've tried to use this module, but so far I always get timeouts. I'm using the following playbook:
``` .yaml
---
- name: RouterOS test
  hosts: router
  gather_facts: no
  tasks:
  - routeros_command:
      commands: /system routerboard print
      host: 192.168.88.1
      port: 22
      transport: routeros
      username: admin
      password: xxx
    register: output
  - debug:
      var: output
```
My inventory:
``` .ini
[router]
router ansible_host=192.168.88.1

[router:vars]
ansible_connection=network_cli
ansible_network_os=routeros
ansible_user=admin
ansible_remote_user=admin
ansible_ssh_pass=xxx
```
When running `ansible-playbook` with `-vvv`, I always get (after waiting for some time):
```
The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/paramiko/channel.py", line 683, in recv
    out = self.in_buffer.read(nbytes, self.timeout)
  File "/usr/lib/python3.6/site-packages/paramiko/buffered_pipe.py", line 160, in read
    raise PipeTimeout()
paramiko.buffered_pipe.PipeTimeout

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../ansible/bin/ansible-connection", line 87, in start
    self.connection._connect()
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 325, in _connect
    newline=self._terminal.terminal_inital_prompt_newline)
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 400, in receive
    data = self._ssh_shell.recv(256)
  File "/usr/lib/python3.6/site-packages/paramiko/channel.py", line 685, in recv
    raise socket.timeout()
socket.timeout

fatal: [router]: FAILED! => {
    "msg": ""
}
```

heuels commented:
@felixfontein Thanks for testing it out! 
The reason you're getting the timeout is because you're trying to login to a shell with console colours enabled. You need to append `+cet` to your username in order for this module to work, like this:
```
routeros:
  hosts:
    example:
      ansible_host: 192.168.88.1
      ansible_user: admin+cet
      ansible_ssh_pass: ''
      ansible_network_os: routeros
      ansible_connection: network_cli
```

I guess I could put it into the docs for this module, however it would be much better if Ansible could add the suffix all by itself. But I'm pretty new to Ansible myself and have not been able to find the way yet.

heuels commented:
I have figured it out. I needed to add the `routeros_command` action plugin whereas I've tried adding `routeros` action plugin. Now the username is replaced properly _and_ automatically.

felixfontein commented:
Yes, adding `+cet` indeed does the trick; with that it works.

I've tried your updated version (with the action plugin), but I still need to append `+cet` for it to work. I've tried debugging this; first, it looks like the action plugin must be called `routeros_command`, otherwise it won't be executed. Next, the SSH connection is opened long before the action plugin is called. The SSH connection is opened by `plugins/connection/network_cli.py` (line 350). Without the `+cet`, the action plugin object isn't even constructed before the connection timeouts.

I've tried tinkering with various things, and I'm not convinced there's anything we can do about it (without deeper changes in `network_cli`). @resmo: do you have any idea? Or know whom to ask about this?

felixfontein commented:
I made a little bit progress: if I copy the following lines from `plugins/terminal/aruba.py`:
``` .py
    ansi_re = [
        # check ECMA-48 Section 5.4 (Control Sequences)
        re.compile(br'(\x1b\[\?1h\x1b=)'),
        re.compile(br'((?:\x9b|\x1b\x5b)[\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e])'),
        re.compile(br'\x08.')
    ]
```
to `plugins/terminal/routeros.py`, I only need to append `+t` ([disable terminal detection](https://wiki.mikrotik.com/wiki/Manual:Console_login_process#Console_login_options)) to the username.
I guess if we could somehow handle the terminal detection, we could survive without `+cet`. In any case, without `+t` or `+cet`, the last thing sent is `'\r\n\r\r\x1bZ  '` (after some banner text).
[... some time later ...]
I can get a bit further, by adding:
``` .py
    terminal_initial_prompt = [
        br'\x1bZ',
    ]
    terminal_initial_answer = b'\x1b/Z'
```
The router then sends `'\x1b<'` and waits again. This apparently means "Enter ANSI mode". I don't know how to react on that. (Also, I don't think `network_cli` allows a proper reaction. Whatever a proper reaction would be...)
So I guess we will have to at least enforce `+t`.

heuels committed:
Handle ANSI codes

heuels commented:
It looks like some kind of caching issue with my setup. I usually test the module against the RouterOS in virtual machine and I actually can somehow log into it without using the `+cet` suffix now. And not only with one username, when I change it it still works. However on a real device it does not, unfortunately.

I have removed the previous commit and tried to strip and/or handle the ANSI codes properly. Turns out, sending space character (` `) is enough to bypass the `'\x1b<'` prompt – at least on my setup 🙂 . I would appreciate it if you could verify it out too.

felixfontein commented:
Yes, now it works! \o/

The only thing that doesn't work is your change to the module itself; (`item = str(item).split('\n')` --> `item = map(lambda x: x.strip(), str(item).split('\n'))`); this causes:
```
Traceback (most recent call last):
  File "/tmp/ansible_or3ract_/ansible_module_routeros_command.py", line 212, in <module>
    main()
  File "/tmp/ansible_or3ract_/ansible_module_routeros_command.py", line 208, in main
    module.exit_json(**result)
  File "/tmp/ansible_or3ract_/ansible_modlib.zip/ansible/module_utils/basic.py", line 2345, in exit_json
  File "/tmp/ansible_or3ract_/ansible_modlib.zip/ansible/module_utils/basic.py", line 2338, in _return_formatted
  File "/tmp/ansible_or3ract_/ansible_modlib.zip/ansible/module_utils/basic.py", line 528, in remove_values
  File "/tmp/ansible_or3ract_/ansible_modlib.zip/ansible/module_utils/basic.py", line 507, in _remove_values_conditions
TypeError: Value of unknown type: <class 'map'>, <map object at 0x7f1883388588>
```
Might be related to me using Python 3. Anyway, that looks rather harmless and easy to fix ;-)

heuels committed:
Revert to_lines function

heuels commented:
Oh well, should have listened to my inner self that insisted on _just fixing the connection issues_ 🙂
Everything should be OK now.

felixfontein commented:
Great! Now it works perfectly fine.

I have Python 3.6.5 as well. I think the problem was that some function in Ansible was processing the result, and didn't like the `map()` return object. If you would have converted it to a list, it would have worked fine. But, see [here](https://docs.quantifiedcode.com/python-anti-patterns/readability/using_map_or_filter_where_list_comprehension_is_possible.html). It is better to do it as follows:
``` .py
    item = [x.strip() for x in str(item).split('\n')]
```
(Besides, it is shorter ;) )

On the other hand, I'm not sure if stripping away spaces is a good idea. Maybe sometimes they will be needed to correctly parse the output.

felixfontein commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,158 @@
+routeros_argument_spec = {}
+
+routeros_top_spec = {
+    'host': dict(removed_in_version=2.9),
```
Are these still necessary? It looks to me (as I said, I'm not familiar with network modules) that this was a now deprecated way to specify the parameters, which you copied from another network module. Since this is a new module, you can probably leave them away.

felixfontein commented on lib/ansible/modules/network/routeros/routeros_command.py:
```diff
@@ -0,0 +1,212 @@
+    required: false
+    default: null
+    aliases: ['waitfor']
+    version_added: "2.2"
```
You can safely remove this line.

felixfontein commented on lib/ansible/modules/network/routeros/routeros_command.py:
```diff
@@ -0,0 +1,212 @@
+    required: false
+    default: all
+    choices: ['any', 'all']
+    version_added: "2.2"
```
Same here.

felixfontein commented on lib/ansible/modules/network/routeros/routeros_command.py:
```diff
@@ -0,0 +1,212 @@
+        is returned. If the I(wait_for) argument is provided, the
+        module is not returned until the condition is satisfied or
+        the number of retries has expired.
+    required: true
```
I *think* it is preferred to use `yes` and `no` in YAML documentation.

heuels replied:
You can actually use [almost whatever you want](http://yaml.org/type/bool.html). Here, I can see that both `required: true` and `require: yes` are used throughout Ansible but the former is used a lot more often. I think I'll keep it.

felixfontein replied:
I know, but I had the impression that there was a tendency to switch from `true`/`false` to `yes`/`no`. But maybe I'm also mistaken.

felixfontein replied:
Let's see if anyone else complains about that when reviewing ;)

felixfontein commented on lib/ansible/plugins/terminal/routeros.py:
```diff
@@ -0,0 +1,72 @@
+    def on_open_shell(self):
+        prompt = self._get_prompt()
+        try:
+            if prompt.strip().endswith(br"\x1b<") or prompt.strip().endswith(b':'):
```
I think you could simply this to `if prompt.strip().endswith(b':'):` here (at least, that works for me), since the first condition is probably never matched. `br"\x1b<"` is the sequence `\`, `x`, `1`, `b`, `<`, while you probably meant `b"\x1b<"`, which is `ESC`, `<`. Anyway, when I change `br"\x1b<"` to `b"\x1b<"`, it doesn't work anymore.

felixfontein replied:
In case you want to know, the error I get when changing `br"\x1b<"` to `b"\x1b<"` is:
```
The full traceback is:
Traceback (most recent call last):
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 243, in exec_command
    cmd = json.loads(to_text(cmd, errors='surrogate_or_strict'))
  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 2 (char 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/paramiko/channel.py", line 683, in recv
    out = self.in_buffer.read(nbytes, self.timeout)
  File "/usr/lib/python3.6/site-packages/paramiko/buffered_pipe.py", line 160, in read
    raise PipeTimeout()
paramiko.buffered_pipe.PipeTimeout

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 437, in send
    response = self.receive(command, prompt, answer, newline, prompt_retry_check)
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 400, in receive
    data = self._ssh_shell.recv(256)
  File "/usr/lib/python3.6/site-packages/paramiko/channel.py", line 685, in recv
    raise socket.timeout()
socket.timeout

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../ansible/lib/ansible/plugins/terminal/routeros.py", line 68, in on_open_shell
    self._exec_cli_command(b' ')
  File ".../ansible/lib/ansible/plugins/terminal/__init__.py", line 73, in _exec_cli_command
    return self._connection.exec_command(cmd)
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 253, in exec_command
    return self.send(command=cmd)
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 441, in send
    raise AnsibleConnectionFailure("timeout trying to send command: %s" % command.strip())
ansible.errors.AnsibleConnectionFailure: timeout trying to send command: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../ansible/bin/ansible-connection", line 87, in start
    self.connection._connect()
  File ".../ansible/lib/ansible/plugins/connection/network_cli.py", line 328, in _connect
    self._terminal.on_open_shell()
  File ".../ansible/lib/ansible/plugins/terminal/routeros.py", line 72, in on_open_shell
    raise AnsibleConnectionFailure('unable to bypass license prompt')
ansible.errors.AnsibleConnectionFailure: unable to bypass license prompt

fatal: [router]: FAILED! => {
    "msg": "unable to bypass license prompt"
}
```

heuels committed:
CR fixes

felixfontein commented:
While reading something about ansible tests, I found [this piece of information](https://docs.ansible.com/ansible/2.3/dev_guide/testing_integration.html#network-tests):
> Starting with Ansible 2.4, all network modules MUST include corresponding unit tests to defend functionality. The unit tests must be added in the same PR that includes the new network module, or extends functionality. Integration tests, although not required, are a welcome addition. How to do this is explained in the rest of this document.

heuels commented:
Sorry, end of the quarter was quite a busy time 🙂

For some reason I'm having a hard time with unit tests. I've copied the contents of `test/units/modules/network/ios` to `test/units/modules/network/routeros` and changed all of the imports and names inside to match. However even the simplest of tests fails like this:
```
~/GitHub/ansible routeros ✗ ◒ → ansible-test units --python 3.6 test/units/modules/network/routeros/test_routeros_command.py
Unit test with Python 3.6
================================= test session starts ==================================
platform darwin -- Python 3.6.5, pytest-3.6.3, py-1.5.4, pluggy-0.6.0
rootdir: /Users/heuels/GitHub/ansible, inifile: tox.ini
plugins: xdist-1.22.2, mock-1.10.0, forked-0.2, f5-sdk-3.0.17
collected 1 item

test/units/modules/network/routeros/test_routeros_command.py F

======================================= FAILURES =======================================
________________ TestRouterosCommandModule.test_routeros_command_simple ________________
self = <units.modules.network.routeros.test_routeros_command.TestRouterosCommandModule testMethod=test_routeros_command_simple>

    def test_routeros_command_simple(self):
        set_module_args(dict(commands=['show version']))
>       result = self.execute_module()

test/units/modules/network/routeros/test_routeros_command.py:64:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/units/modules/network/routeros/routeros_module.py:60: in execute_module
    result = self.changed(changed)
test/units/modules/network/routeros/routeros_module.py:81: in changed
    self.module.main()
lib/ansible/modules/network/routeros/routeros_command.py:180: in main
    responses = run_commands(module, module.params['commands'])
../../.pyenv/versions/3.6.5/lib/python3.6/unittest/mock.py:939: in __call__
    return _mock_self._mock_call(*args, **kwargs)
../../.pyenv/versions/3.6.5/lib/python3.6/unittest/mock.py:1005: in _mock_call
    ret_val = effect(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def load_from_file(*args, **kwargs):
        module, commands = args
        output = list()

        for item in commands:
            try:
>               obj = json.loads(item['command'])
E               TypeError: string indices must be integers

test/units/modules/network/routeros/test_routeros_command.py:52: TypeError
 generated xml file: /Users/heuels/GitHub/ansible/test/results/junit/python3.6-units.xml
=============================== short test summary info ================================
FAIL test/units/modules/network/routeros/test_routeros_command.py::TestRouterosCommandModule::test_routeros_command_simple
=============================== 1 failed in 0.29 seconds ===============================
ERROR: Command "pytest --boxed -r a --color yes --junit-xml test/results/junit/python3.6-units.xml test/units/modules/network/routeros/test_routeros_command.py" returned exit status 1.
```

Obviously, running `test_ios_command.py` succeeds. I feel like I'm missing something important 🙂

felixfontein commented:
These errors appear because the `ios_command` module internally works differently: it converts its `commands` list into a list of dictionaries (which have a key `command`, which is the original string). So in the test, you have to replace `item['command']` by `item` itself.

The next step is probably changing the tests to use valid RouterOS commands, and creating fitting fixture files in `test/units/modules/network/routeros/fixtures`.

felixfontein commented:
@heuels can you put your test trials into a (separate) branch of https://github.com/heuels/ansible? Then I can also play around with it without doing duplicate work ;-)

heuels committed:
test(routeros): add unit tests

heuels commented:
@felixfontein, Sure! I've added unit tests to [`routeros-units`](https://github.com/heuels/ansible/compare/routeros...heuels:routeros-units) branch of my fork. Can merge them into this branch whenever you're ready 🙂

felixfontein commented:
@heuels You can also merge them here, if you want, though that will probably break the tests ;-) (or did you get them working?)

heuels commented:
Yes, all the tests pass. That's kinda the idea, right? 🙂

felixfontein commented:
Yes, it is! ;-)

felixfontein committed:
Added another test (with ANSI colors and banner in fixture).

felixfontein commented:
@heuels I added another test in heuels/ansible#1; this one uses fancy stuff such as ANSI colors ;-) Feel free to change it as you want.

heuels commented:
@felixfontein Merged it, thanks for the PR!

felixfontein commented:
Hmm, I think the sanity checker should ignore fixture files. The `\r\n` line endings is what the real hardware is sending. @mattclay would it be possible to ignore fixtures for the line endings test?

mattclay commented:
The line ending test can be skipped for specific files as needed by adding them here:

https://github.com/ansible/ansible/blob/devel/test/sanity/code-smell/line-endings.py#L13

felixfontein commented:
@mattclay Thanks!

@heuels What do you think, do you want to return to the old fixtures with `\r\n` and add it to the `line-endings.py` script? I think it would be a good idea, so that failling unit tests start screaming when something's broken with line ending parsing...

heuels commented:
@felixfontein Yes, I think this will be the right way to do it.

heuels committed:
Ignore CRLF line endings in system_package_print file

felixfontein reviewed (approved):
I think it's good to go :)

felixfontein commented:
shipit

heuels commented:
@felixfontein Thanks! And thanks once more for your invaluable help with this PR!

Just one more thing I noticed tonight: if a command is somewhat longish it gets truncated by the RouterOS console and this leads to parts of command itself appearing in the output. 

Do you think it would be a good idea to remove from the output lines that start either with `'/'` (commands), `'>'` (truncated parts of commands) and empty strings?

Here is an example output:
```
ok: [example.com] => {
    "changed": false,
    "invocation": {
        "module_args": {
            "commands": [
                "/interface wireless print count-only where interface-type!=virtual"
            ],
            "interval": 1,
            "match": "all",
            "retries": 10,
            "wait_for": null
        }
    },
    "stdout": [
        "/interface wireless print count-only where interface-type!=v\n<rint count-only where interface-type!=vi                                      \n<rint count-only where interface-type!=virtual\n<rint count-only where interface-type!=virtual\n\n0"
    ],
    "stdout_lines": [
        [
            "/interface wireless print count-only where interface-type!=v",
            "<rint count-only where interface-type!=vi                                      ",
            "<rint count-only where interface-type!=virtual",
            "<rint count-only where interface-type!=virtual",
            "",
            "0"
        ]
    ]
}
```

felixfontein commented:
@heuels Hmm, this is complicated. Removing things from output can be dangerous, think about if you're trying to dump a script or log and suddenly parts of it get removed. I don't know enough about RouterOS to say whether that's indeed a problem or not...
Maybe it is possible to tell RouterOS that our "terminal" has more columns, so it stops doing that. No idea how to do that though...

heuels commented:
@felixfontein It is possible, though hacky as always. One would need to manually add `+t1000w` to their username (where 1000 is a number of columns). It's a pity this behaviour is too OS-specific, otherwise we could consider implementing some kind of `username_suffix` that can be controlled from the module itself.

felixfontein commented:
@heuels There should also be other ways to do this; after all, terminals can inform over a SSH session that their size changed. But I don't really know how they do that...
Anyway, it would be nicer if there's some API which allows to send commands and receive results *without* all the terminal crap. In fact, having structured output (like JSON, YAML, ...) would be even nicer ;-)

felixfontein commented:
@heuels Aaah, yes, I think I've seen that already somewhere. I guess that would simplify everything. But yes, it is definitely out of scope for this PR IMO.

ganeshrn commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,148 @@
+
+_DEVICE_CONFIGS = {}
+
+routeros_provider_spec = {
```
Provider spec is required to support connection=local type and is present in other network platform for legacy reason, in future local connection type for network module will be deprecated. 
As this platform is newly added supporting local connection type is not required imo and this spec can be removed.

heuels replied:
I am planning on implementing support for RouterOS API as well. I could implement it either with a provider dict, or by creating a special connection type. Which do you think would be better?

RouterOS has a weird non-standard API, so it seems that it’s not possible to use any of the existing connection types, except for local, of course.

ganeshrn replied:
Which transport type is used by RouterOS API? 
You can check [httapi](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/httpapi.py) connection type and implementation plugins for [nxap/eapi](https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/httpapi).

heuels replied:
It uses socket connection with custom login process, not HTTP(S) unfortunately.
There's a library called [librouteros](https://github.com/luqasz/librouteros) that implements RouterOS API in Python, here is how it creates a connection: https://github.com/luqasz/librouteros/blob/83fa287fa9a12d209687685f3c77eb80d1dd32f3/librouteros/__init__.py#L63-L70

ganeshrn replied:
In that case, it seems connection=local is the way forward for now. You can ignore my comment on providers.

ganeshrn commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,148 @@
+routeros_argument_spec = {}
+
+
+def get_provider_argspec():
```
`get_provider_argspec` again added to support local connection type with platform agnostic modules. This function can be removed.

ganeshrn commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,148 @@
+    return module._routeros_capabilities
+
+
+def check_args(module, warnings):
```
`check_args` is present in other network platforms for legacy reasons and can be removed here if not used

ganeshrn commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,148 @@
+
+def get_defaults_flag(module):
+    connection = get_connection(module)
+    out = connection.get('/system default-configuration print')
```
This call on connection object can raise ConnectionError exception which needs to handled here else it will result in stack trace in output.  I have raised a https://github.com/ansible/ansible/pull/43353 to fix it for other supported platforms

ganeshrn commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,148 @@
+        return _DEVICE_CONFIGS[flag_str]
+    except KeyError:
+        connection = get_connection(module)
+        out = connection.get_config(flags=flags)
```
same as above

ganeshrn commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,148 @@
+            prompt = None
+            answer = None
+
+        out = connection.get(command, prompt, answer)
```
same as above

ganeshrn commented on lib/ansible/module_utils/network/routeros/routeros.py:
```diff
@@ -0,0 +1,148 @@
+
+        try:
+            out = to_text(out, errors='surrogate_or_strict')
+            print("run_commands: %s" % str(out))
```
print statement can be removed

ganeshrn commented on lib/ansible/modules/network/routeros/routeros_command.py:
```diff
@@ -0,0 +1,210 @@
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'network'}
```
This should be `'supported_by': 'community'`

ganeshrn commented on lib/ansible/modules/network/routeros/routeros_command.py:
```diff
@@ -0,0 +1,210 @@
+        See examples.
+    required: false
+    default: null
+    aliases: ['waitfor']
```
`waitfor` alias not required, added in other platform for legacy reasons

ganeshrn reviewed (commented):
Please add your Github handle as maintainer for the new files added in https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml

ganeshrn reviewed (commented):
Also please documentation for this platform in https://github.com/ansible/ansible/tree/devel/docs/docsite/rst/network/user_guide
https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/network/user_guide/platform_index.rst

heuels committed:
fix: review by ganeshrn

felixfontein commented on docs/docsite/rst/network/user_guide/platform_routeros.rst:
```diff
@@ -0,0 +1,65 @@
+Using CLI in Ansible 2.6
+================================================================================
+
+Example CLI ``group_vars/ios.yml``
```
I guess you should change `ios.yml` to `routeros.yml` ;-)

heuels replied:
My bad, thanks! 🙂

heuels commented:
@ganeshrn I've addressed all your comments, thank you for reviewing!

ganeshrn commented:
@heuels @felixfontein  Thank you!

felixfontein commented:
@ganeshrn Thank you very much for reviewing and merging this!
@heuels And thank you very much for all your work for this!
I'm really looking forward to use this :)
